src/commands/subcommand.cc

```diff
@@ -144,37 +136,78 @@ namespace	// subcommand detetction
     return false;
   }
 
-  template <class OutputIterator_>
-  unsigned collectSubommandsIn( const Pathname & dir_r, OutputIterator_ result_r )
+  /** Return whether \a dir_r/name_r forms a valid subcommand. */
+  inline SubcommandOptions::Detected detectSubcommand( const Pathname & dir_r, std::string name_r )
   {
-    unsigned cnt = 0;
+    SubcommandOptions::Detected ret;
+    if ( str::startsWith( name_r, "zypper-" ) && canExecute( dir_r/name_r ) ) {
+      ret._cmd  = name_r.substr( 7 /*"zypper-"*/ );
+      ret._name = std::move(name_r);
+      ret._path = dir_r;
+    }
+    return ret;
+  }
+
+  /** Collect subcommands found in \a dir_r. */
+  inline void detectSubcommandsIn( const Pathname & dir_r, std::function<void(SubcommandOptions::Detected)> fnc_r )
+  {
+    if ( !fnc_r )
+      return;
+
     filesystem::dirForEach( dir_r,
-			    [&result_r]( const Pathname & dir_r, std::string str_r )->bool
+			    [&fnc_r]( const Pathname & dir_r, std::string name_r )->bool
 			    {
-			      if ( str::startsWith( str_r, "zypper-" ) && canExecute( dir_r/str_r ) )
-			      {
-				str_r.erase( 0, 7 /*"zypper-"*/ );
-				*result_r = str_r;
-			      }
+			      SubcommandOptions::Detected cmd { detectSubcommand( dir_r, std::move(name_r) ) };
+			      if ( ! cmd._cmd.empty() )
+				fnc_r( std::move(cmd) );
 			      return true;
 			    } );
-    return cnt;
   }
 
-  inline void collectAllSubommands( std::set<std::string> & execdirCommands_r, std::set<std::string> & pathCommands_r )
+  /* Just the command names for the short help. */
+  inline void collectAllSubommands( std::set<std::string> & allCommands_r )
   {
-    collectSubommandsIn( SubcommandOptions::_execdir, std::inserter(execdirCommands_r,execdirCommands_r.end()) );
+    auto collectSubommandsIn = [&allCommands_r]( const Pathname & dir_r ) {
+      detectSubcommandsIn( dir_r,
+			   [&allCommands_r]( SubcommandOptions::Detected cmd_r )
+			   {
+			     allCommands_r.insert( std::move(cmd_r._cmd) );
+			   } );
+    };
+
+    collectSubommandsIn( SubcommandOptions::_execdir );
 
     std::set<Pathname> dirs;
     str::split( env::PATH(), std::inserter(dirs,dirs.end()), ":" );
     for ( const auto & dir : dirs )
     {
-      collectSubommandsIn( dir, std::inserter(pathCommands_r,pathCommands_r.end()) );
+      collectSubommandsIn( dir );
     }
   }
 
-  inline void collectAllSubommands( std::set<std::string> & allCommands_r )
-  { collectAllSubommands( allCommands_r, allCommands_r ); }
+  /* The command details for the long help. */
```

Comment:
Does this comment belong here?

Reply:
Yes. `/* Just the command names for the short help. */` vs `/* The command details for the long help. */`.